File: kod-ornekleri/java/TodoApp.java
/**
 * Java Örnek: Todo Uygulaması
 * Yazılım Geliştirme Araçları Eğitimi - Java Örneği
 * 
 * Bu örnek şunları gösterir:
 * - Java class yapısı
 * - ArrayList kullanımı
 * - Exception handling
 * - JavaDoc yorumları
 */

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class TodoApp {
    private List<Todo> todos;
    private int nextId;

    /**
     * Constructor - TodoApp nesnesini başlatır
     */
    public TodoApp() {
        this.todos = new ArrayList<>();
        this.nextId = 1;
    }

    /**
     * Yeni görev ekler
     * 
     * @param title Görev başlığı
     * @return Eklenen Todo nesnesi
     * @throws IllegalArgumentException Başlık boş ise
     */
    public Todo addTodo(String title) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Görev başlığı boş olamaz");
        }

        Todo newTodo = new Todo(nextId++, title.trim());
        todos.add(newTodo);
        return newTodo;
    }

    /**
     * Görevi tamamlandı olarak işaretler
     * 
     * @param id Görev ID
     * @return İşlem başarılı ise true
     */
    public boolean toggleTodo(int id) {
        Todo todo = findTodoById(id);
        if (todo != null) {
            todo.setCompleted(!todo.isCompleted());
            return true;
        }
        return false;
    }

    /**
     * Görevi siler
     * 
     * @param id Görev ID
     * @return İşlem başarılı ise true
     */
    public boolean deleteTodo(int id) {
        return todos.removeIf(todo -> todo.getId() == id);
    }

    /**
     * Tüm görevleri getirir
     * 
     * @return Görev listesi
     */
    public List<Todo> getTodos() {
        return new ArrayList<>(todos);
    }

    /**
     * Tamamlanan görevleri getirir
     * 
     * @return Tamamlanan görev listesi
     */
    public List<Todo> getCompletedTodos() {
        return todos.stream()
                .filter(Todo::isCompleted)
                .collect(Collectors.toList());
    }

    /**
     * Aktif görevleri getirir
     * 
     * @return Aktif görev listesi
     */
    public List<Todo> getActiveTodos() {
        return todos.stream()
                .filter(todo -> !todo.isCompleted())
                .collect(Collectors.toList());
    }

    /**
     * ID'ye göre görev bulur
     * 
     * @param id Görev ID
     * @return Bulunan Todo nesnesi veya null
     */
    private Todo findTodoById(int id) {
        return todos.stream()
                .filter(todo -> todo.getId() == id)
                .findFirst()
                .orElse(null);
    }

    /**
     * Tüm görevleri temizler
     */
    public void clearAll() {
        todos.clear();
        nextId = 1;
    }

    /**
     * Ana metod - Örnek kullanım
     */
    public static void main(String[] args) {
        TodoApp app = new TodoApp();
        
        // Örnek görevler ekle
        app.addTodo("Java öğren");
        app.addTodo("Test yaz");
        
        // Görevleri listele
        System.out.println("Tüm görevler:");
        app.getTodos().forEach(System.out::println);
        
        // İlk görevi tamamla
        app.toggleTodo(1);
        
        System.out.println("\nTamamlanan görevler:");
        app.getCompletedTodos().forEach(System.out::println);
    }
}

/**
 * Todo sınıfı - Görev bilgilerini tutar
 */
class Todo {
    private int id;
    private String title;
    private boolean completed;
    private String createdAt;

    public Todo(int id, String title) {
        this.id = id;
        this.title = title;
        this.completed = false;
        this.createdAt = java.time.LocalDateTime.now().toString();
    }

    // Getter ve Setter metodları
    public int getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public boolean isCompleted() {
        return completed;
    }

    public void setCompleted(boolean completed) {
        this.completed = completed;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    @Override
    public String toString() {
        return String.format("Todo{id=%d, title='%s', completed=%s}", 
                id, title, completed);
    }
}

File: kod-ornekleri/csharp/TodoApp.cs
/**
 * C# Örnek: Todo Uygulaması
 * Yazılım Geliştirme Araçları Eğitimi - C# Örneği
 * 
 * Bu örnek şunları gösterir:
 * - C# class yapısı
 * - List<T> kullanımı
 * - LINQ sorguları
 * - Properties
 * - Exception handling
 */

using System;
using System.Collections.Generic;
using System.Linq;

namespace TodoAppExample
{
    /// <summary>
    /// Todo sınıfı - Görev bilgilerini tutar
    /// </summary>
    public class Todo
    {
        public int Id { get; set; }
        public string Title { get; set; }
        public bool Completed { get; set; }
        public DateTime CreatedAt { get; set; }

        public Todo(int id, string title)
        {
            Id = id;
            Title = title;
            Completed = false;
            CreatedAt = DateTime.Now;
        }

        public override string ToString()
        {
            return $"Todo{{Id={Id}, Title='{Title}', Completed={Completed}}}";
        }
    }

    /// <summary>
    /// TodoApp sınıfı - Görev yönetimi işlemlerini yapar
    /// </summary>
    public class TodoApp
    {
        private List<Todo> todos;
        private int nextId;

        /// <summary>
        /// Constructor - TodoApp nesnesini başlatır
        /// </summary>
        public TodoApp()
        {
            todos = new List<Todo>();
            nextId = 1;
        }

        /// <summary>
        /// Yeni görev ekler
        /// </summary>
        /// <param name="title">Görev başlığı</param>
        /// <returns>Eklenen Todo nesnesi</returns>
        /// <exception cref="ArgumentException">Başlık boş ise</exception>
        public Todo AddTodo(string title)
        {
            if (string.IsNullOrWhiteSpace(title))
            {
                throw new ArgumentException("Görev başlığı boş olamaz", nameof(title));
            }

            var newTodo = new Todo(nextId++, title.Trim());
            todos.Add(newTodo);
            return newTodo;
        }

        /// <summary>
        /// Görevi tamamlandı olarak işaretler
        /// </summary>
        /// <param name="id">Görev ID</param>
        /// <returns>İşlem başarılı ise true</returns>
        public bool ToggleTodo(int id)
        {
            var todo = FindTodoById(id);
            if (todo != null)
            {
                todo.Completed = !todo.Completed;
                return true;
            }
            return false;
        }

        /// <summary>
        /// Görevi siler
        /// </summary>
        /// <param name="id">Görev ID</param>
        /// <returns>İşlem başarılı ise true</returns>
        public bool DeleteTodo(int id)
        {
            var todo = FindTodoById(id);
            if (todo != null)
            {
                todos.Remove(todo);
                return true;
            }
            return false;
        }

        /// <summary>
        /// Tüm görevleri getirir
        /// </summary>
        /// <returns>Görev listesi</returns>
        public List<Todo> GetTodos()
        {
            return new List<Todo>(todos);
        }

        /// <summary>
        /// Tamamlanan görevleri getirir
        /// </summary>
        /// <returns>Tamamlanan görev listesi</returns>
        public List<Todo> GetCompletedTodos()
        {
            return todos.Where(t => t.Completed).ToList();
        }

        /// <summary>
        /// Aktif görevleri getirir
        /// </summary>
        /// <returns>Aktif görev listesi</returns>
        public List<Todo> GetActiveTodos()
        {
            return todos.Where(t => !t.Completed).ToList();
        }

        /// <summary>
        /// ID'ye göre görev bulur
        /// </summary>
        /// <param name="id">Görev ID</param>
        /// <returns>Bulunan Todo nesnesi veya null</returns>
        private Todo FindTodoById(int id)
        {
            return todos.FirstOrDefault(t => t.Id == id);
        }

        /// <summary>
        /// Tüm görevleri temizler
        /// </summary>
        public void ClearAll()
        {
            todos.Clear();
            nextId = 1;
        }

        /// <summary>
        /// Ana metod - Örnek kullanım
        /// </summary>
        public static void Main(string[] args)
        {
            var app = new TodoApp();

            // Örnek görevler ekle
            app.AddTodo("C# öğren");
            app.AddTodo("Test yaz");

            // Görevleri listele
            Console.WriteLine("Tüm görevler:");
            foreach (var todo in app.GetTodos())
            {
                Console.WriteLine(todo);
            }

            // İlk görevi tamamla
            app.ToggleTodo(1);

            Console.WriteLine("\nTamamlanan görevler:");
            foreach (var todo in app.GetCompletedTodos())
            {
                Console.WriteLine(todo);
            }
        }
    }
}

